File: zadatak1/zadatak.py
class DFSClass:
	def printOptions(self, s):
		#print(s)
		#prva zagrada
		try:
			ind = s.index('{')
		except:
	        		ind=-1

		if (ind<0): #nema zagrada
			print(s)
		else:
			#naci odgovarajucu }
			nestCount=0
			endIndex=ind
			for i in range(ind+1,len(s)):
				if (s[i]=='{'):
					nestCount+=1
				elif (s[i]=='}'):
					if (nestCount==0):
						endIndex=i
						break
					nestCount-=1

			if (endIndex==ind):
				# nevalja format, nema }
				raise Exception('Invalid format!')

			# naci opcije u zagradi
			options = []
				
			startIndex=ind+1
			nestCount=0
				
			for i in range(startIndex,endIndex):
				if (s[i]=='{'):
					nestCount+=1
				elif (s[i]=='}'):
					nestCount-=1
				elif (s[i]=='|'):
					if (nestCount==0):
						option = s[startIndex:i]
						options.append(option)
						startIndex = i+1
		
			if (startIndex<endIndex-1): #zadnja opcija, ostatak prije }
				option = s[startIndex:endIndex]
				options.append(option)
		
			prefix = ""
			if (ind>0):
				prefix = s[0:ind]
				
			sufix = ""
			if (endIndex<len(s)-1):
				sufix = s[endIndex+1:len(s)]

			#rekurzija, isti string bez prve zagrade
			for option in options:
				self.printOptions(prefix+option+sufix)

result = DFSClass()
str = u'{ Danas pada { kisa { i ulice su mokre } | snijeg { i ulice su { klizave | bijele }}}. Sutra ce biti lijepo { vrijeme | setati }. }'
result.printOptions(str)

File: zadatak1/zadatak.java
import java.lang.Math;
import java.util.List;
import java.util.ArrayList;

public class MainClass {
  public static void main(String[] args)throws Exception {
  	new DFSClass().printOptions("{ Danas pada { kisa { i ulice su mokre } | snijeg { i ulice su { klizave | bijele |crne}}}. Sutra ce biti lijepo { vrijeme | setati }. }{Ovo je moj {primjer|zadatak}}");
  }
}

public class DFSClass {
  public void printOptions(String s)throws Exception{

	//prva zagrada
	int ind = s.indexOf('{');

	if (ind<0){ //nema zagrada
		System.out.println(s);
	}else{
		//naci odgovarajucu }
		int nestCount=0;
		int endIndex=ind;
		for (int i=ind+1;i<s.length();i++){
			if (s.charAt(i)=='{'){
				nestCount++;
			}else if (s.charAt(i)=='}'){
				if (nestCount==0){
					endIndex=i;
					break;
				}
				nestCount--;
			}
		}
		if (endIndex==ind){
			// nevalja format, nema }
			throw new Exception("invalid format");
		}

		// naci opcije u zagradi
		List<String> options = new ArrayList<String>();
		
        int startIndex=ind+1;
		
        nestCount=0;
		
        for (int i=startIndex; i<endIndex;i++){
			if (s.charAt(i)=='{'){
				nestCount++;
			}else if (s.charAt(i)=='}'){
				nestCount--;
			}else if (s.charAt(i)=='|'){
				if (nestCount==0){
					String option = s.substring(startIndex,i);
					options.add(option);
					startIndex = i+1;
				}
			}
		}
		
        if (startIndex<endIndex-1){ //zadnja opcija, ostatak prije }
			String option = s.substring(startIndex,endIndex);
			options.add(option);
		}

		String prefix = "";
		if (ind>0){
			prefix = s.substring(0,ind);
		}
		
        String sufix = "";
		if (endIndex<s.length()-1){
			sufix = s.substring(endIndex+1,s.length());
		}
		
        //rekurzija, isti string bez prve zagrade
		for (String option : options){
			printOptions(prefix+option+sufix);
		}
	}
}
}
